services/src/LinkLossService/LinkLossService.cpp

```diff
@@ -0,0 +1,110 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mbed-os-experimental-ble-services/services/inc/LinkLossService.h"
+
+LinkLossService::LinkLossService(BLE &ble, events::EventQueue &event_queue, ChainableGapEventHandler &chainable_gap_event_handler) :
+    _ble(ble),
+    _event_queue(event_queue),
+    _chainable_gap_event_handler(chainable_gap_event_handler),
+    _alert_level_char(GattCharacteristic::UUID_ALERT_LEVEL_CHAR, &_alert_level)
+{
+}
+
+LinkLossService::~LinkLossService()
+{
+}
+
+ble_error_t LinkLossService::init()
+{
+    GattCharacteristic *charTable[] = { &_alert_level_char };
+    GattService         linkLossService(GattService::UUID_LINK_LOSS_SERVICE, charTable, 1);
+
+    _alert_level_char.setWriteAuthorizationCallback(this, &LinkLossService::onDataWritten);
+
+    ble_error_t bleError = _ble.gattServer().addService(linkLossService);
+
+    if (!bleError) {
+        _chainable_gap_event_handler.addEventHandler(this);
+    }
+
+    return bleError;
+}
+
+void LinkLossService::set_event_handler(EventHandler* handler)
+{
+    alert_handler = handler;
+}
+
+void LinkLossService::set_level(AlertLevel level)
+{
+    _alert_level = level;
+}
+
+void LinkLossService::set_timeout(std::chrono::milliseconds timeout)
+{
+    _alert_timeout = timeout;
+}
+
+LinkLossService::AlertLevel LinkLossService::get_level()
+{
+    return _alert_level;
+}
+
+void LinkLossService::stop_alert()
+{
+    _event_queue.cancel(_event_queue_handle);
+    if (in_alert) {
+        in_alert = false;
+        if (alert_handler) {
+            alert_handler->on_alert_end();
+        }
+    }
+}
+
+void LinkLossService::onConnectionComplete(const ble::ConnectionCompleteEvent &event)
+{
+    stop_alert();
+}
+
+void LinkLossService::onDisconnectionComplete(const ble::DisconnectionCompleteEvent &event)
+{
+    AlertLevel level = get_level();
+    if (event.getReason() == ble::disconnection_reason_t::CONNECTION_TIMEOUT && level != AlertLevel::NO_ALERT) {
+        in_alert = true;
+        alert_handler->on_alert_requested(level);
+        if (_alert_timeout > std::chrono::milliseconds(0)) {
+            _event_queue_handle = _event_queue.call_in(_alert_timeout, [this] { stop_alert(); });
+        }
+    }
+}
+
+void LinkLossService::onDataWritten(GattWriteAuthCallbackParams *write_request)
+{
+    const uint8_t level = *write_request->data;
+
+    if (write_request->offset != 0) {
+        write_request->authorizationReply = AUTH_CALLBACK_REPLY_ATTERR_INVALID_OFFSET;
+        return;
+    }
+
+    if (level <= *reinterpret_cast<uint8_t *>(AlertLevel::HIGH_ALERT)) {
+        set_level((AlertLevel) level);
+    } else {
+        write_request->authorizationReply = AUTH_CALLBACK_REPLY_ATTERR_WRITE_NOT_PERMITTED;
+    }
```

Comment:
if the BLE device is compliant it cannot write anything else in the service, maybe we can dispense with the checking (you'd still need to check the value before the enum cast)